File: go/api/api_credits.go
// api/api_credits.go
package api

import (
	"github.com/weft-finance/vayu-go/client"
	"github.com/weft-finance/vayu-go/openapi"
)

type CreditsAPI struct {
	vayuClient *client.VayuClient
}

type DeductCreditsRequest = openapi.DeductCreditsRequest
type GrantCreditsRequest = openapi.GrantCreditsRequest
type ListCreditLedgerEntriesResponse = openapi.ListCreditLedgerEntriesResponse

func NewCreditsAPI(client *client.VayuClient) *CreditsAPI {
	return &CreditsAPI{
		vayuClient: client,
	}
}

func (api *CreditsAPI) DeductCredits(input DeductCreditsRequest) error {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CreditsAPI.DeductCredits(ctx)
	request = request.DeductCreditsRequest((openapi.DeductCreditsRequest)(input))
	_, err := request.Execute()

	if err != nil {
		return client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return nil
}

func (api *CreditsAPI) GrantCredits(input GrantCreditsRequest) error {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CreditsAPI.GrantCredits(ctx)
	request = request.GrantCreditsRequest((openapi.GrantCreditsRequest)(input))
	_, err := request.Execute()

	if err != nil {
		return client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return nil
}

func (api *CreditsAPI) ListCreditLedgerEntries(customerId string, limit *float32, cursor *string) (*ListCreditLedgerEntriesResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CreditsAPI.ListCreditLedgerEntries(ctx)
	request = request.CustomerId(customerId)

	if limit != nil {
		request = request.Limit(*limit)
	}

	if cursor != nil {
		request = request.Cursor(*cursor)
	}

	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

File: python/vayu_sdk/apis/api_credits.py
# vayu_sdk/apis/api_credits.py
from openapi.api.credits_api import CreditsApi
from vayu_sdk.clients.vayu_client import VayuClient
from openapi.models.deduct_credits_request import DeductCreditsRequest
from openapi.models.grant_credits_request import GrantCreditsRequest
from openapi.models.list_credit_ledger_entries_response import ListCreditLedgerEntriesResponse


class CreditsAPI:
    __client: CreditsApi = None

    def __init__(self, vayu_client: VayuClient):
        self.__client = CreditsApi(vayu_client.client)

    def deduct(self, request: DeductCreditsRequest):
        return self.__client.deduct_credits(deduct_credits_request=request)

    def grant(self, request: GrantCreditsRequest):
        return self.__client.grant_credits(grant_credits_request=request)

    def list_ledger_entries(self, customer_id: str, limit: int = None, cursor: str = None):
        return self.__client.list_credit_ledger_entries(
            customer_id=customer_id, limit=limit, cursor=cursor
        )
